### ISSUE TITLE
j2objc-0.9.6: Java Serialization of Classes Fails

### ISSUE THREAD

**Issue by mvoong** [google/j2objc#496]
Serializing an object works fine, but a class fails in all of these example cases. Output provided at the end.

```java
import java.io.*;

public class Tests {
    public static void performTest() {
        try {
            testSerializingObjectClass();
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            testSerializingSerializableClass();
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            testSerializingNotSerializableClass();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    public static void testSerializingObjectClass() throws Exception {
        String path = "/tmp/a.object";
        FileOutputStream fileOut = new FileOutputStream(path);
        new ObjectOutputStream(fileOut).writeObject(Object.class);
        FileInputStream fileIn = new FileInputStream(path);
        new ObjectInputStream(fileIn).readObject();
    }
    
    public static void testSerializingSerializableClass() throws Exception {
        String path = "/tmp/b.object";
        FileOutputStream fileOut = new FileOutputStream(path);
        new ObjectOutputStream(fileOut).writeObject(SerializableClass.class);
        FileInputStream fileIn = new FileInputStream(path);
        new ObjectInputStream(fileIn).readObject();
    }
    
    public static void testSerializingNotSerializableClass() throws Exception {
        String path = "/tmp/c.object";
        FileOutputStream fileOut = new FileOutputStream(path);
        new ObjectOutputStream(fileOut).writeObject(NotSerializableClass.class);
        FileInputStream fileIn = new FileInputStream(path);
        new ObjectInputStream(fileIn).readObject();
    }

    static class SerializableClass implements Serializable {
        
    }
    
    static class NotSerializableClass {
        
    }
}
```

Output:
```
java.io.NotSerializableException: java.lang.Object
	at 0x000000010a47b109 java.lang.Exception.<init>() + 39
	at 0x000000010a44941e java.io.IOException.<init>() + 39
	at 0x000000010a458a03 java.io.ObjectStreamException.<init>() + 39
	at 0x000000010a449619 java.io.NotSerializableException.<init>() + 39
	at 0x000000010a45345b java.io.ObjectOutputStream.writeNewObject() + 1029
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a4508f5 java.io.ObjectOutputStream.writeFieldValues() + 1023
	at 0x000000010a45204e java.io.ObjectOutputStream.writeHierarchy() + 469
	at 0x000000010a4532c1 java.io.ObjectOutputStream.writeNewObject() + 619
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a433848 Tests.testSerializingObjectClass() + 168
	at 0x000000010a4335dd Tests.performTest() + 13
	at 0x000000010a433505 Tests.performTest() + 21

java.io.NotSerializableException: Protocol
	at 0x000000010a47b109 java.lang.Exception.<init>() + 39
	at 0x000000010a44941e java.io.IOException.<init>() + 39
	at 0x000000010a458a03 java.io.ObjectStreamException.<init>() + 39
	at 0x000000010a449619 java.io.NotSerializableException.<init>() + 39
	at 0x000000010a45345b java.io.ObjectOutputStream.writeNewObject() + 1029
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a4508f5 java.io.ObjectOutputStream.writeFieldValues() + 1023
	at 0x000000010a45204e java.io.ObjectOutputStream.writeHierarchy() + 469
	at 0x000000010a4532c1 java.io.ObjectOutputStream.writeNewObject() + 619
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a45272a java.io.ObjectOutputStream.writeNewArray() + 717
	at 0x000000010a453ce1 java.io.ObjectOutputStream.writeObjectInternal() + 989
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a4508f5 java.io.ObjectOutputStream.writeFieldValues() + 1023
	at 0x000000010a45204e java.io.ObjectOutputStream.writeHierarchy() + 469
	at 0x000000010a4532c1 java.io.ObjectOutputStream.writeNewObject() + 619
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a433986 Tests.testSerializingSerializableClass() + 166
	at 0x000000010a4335ec Tests.performTest() + 28
	at 0x000000010a433505 Tests.performTest() + 21

java.io.NotSerializableException: Tests$NotSerializableClass
	at 0x000000010a47b109 java.lang.Exception.<init>() + 39
	at 0x000000010a44941e java.io.IOException.<init>() + 39
	at 0x000000010a458a03 java.io.ObjectStreamException.<init>() + 39
	at 0x000000010a449619 java.io.NotSerializableException.<init>() + 39
	at 0x000000010a45345b java.io.ObjectOutputStream.writeNewObject() + 1029
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a4508f5 java.io.ObjectOutputStream.writeFieldValues() + 1023
	at 0x000000010a45204e java.io.ObjectOutputStream.writeHierarchy() + 469
	at 0x000000010a4532c1 java.io.ObjectOutputStream.writeNewObject() + 619
	at 0x000000010a453e46 java.io.ObjectOutputStream.writeObjectInternal() + 1346
	at 0x000000010a45382a java.io.ObjectOutputStream.writeObject() + 199
	at 0x000000010a433ac6 Tests.testSerializingNotSerializableClass() + 166
	at 0x000000010a4335fb Tests.performTest() + 43
	at 0x000000010a433505 Tests.performTest() + 21
```

**Comment by nbransby** [comment]
Any tips on this one @tomball? We need this working soon so happy to attempt the fix with some pointers

**Comment by tomball** [comment]
It hasn't been reviewed by Keith yet, but here's a patch that should work. I was going to send pointers as you suggested, but when the problem wandered into weak-reference support, I thought it best to just buckle down and fix it. The patch looks more verbose than it really is -- you'll see that you wrote most of it. :-)

To apply this patch, copy the following into a temp file, then in your j2objc root dir, run "patch -p2 < temp-file".
```
diff --git a/j2objc/jre_emul/Classes/IOSClass.m b/j2objc/jre_emul/Classes/IOSClass.m
index ae72fd1..8e22110 100644
--- a/j2objc/jre_emul/Classes/IOSClass.m
+++ b/j2objc/jre_emul/Classes/IOSClass.m
@@ -879,6 +879,10 @@ IOSObjectArray *copyFieldsToObjectArray(NSArray *fields) {
   return self;
 }

+- (IOSClass *)getClass {
+  return IOSClass_class_();
+}
+
 static BOOL IsStringType(Class cls) {
   // We can't trigger class initialization because that might recursively enter
   // FetchClass and result in deadlock within the FastPointerLookup. Therefore,
diff --git a/j2objc/jre_emul/Classes/IOSReference.m b/j2objc/jre_emul/Classes/IOSReference.m
index 4374bb4..a45c2b1 100644
--- a/j2objc/jre_emul/Classes/IOSReference.m
+++ b/j2objc/jre_emul/Classes/IOSReference.m
@@ -176,6 +176,10 @@ static Class GetReferentSubclass(id obj) {
 // (since they are never GC'd), with this test they will do nothing in
 // iOS as well.
 static BOOL IsConstantObject(id obj) {
+  if ([obj isKindOfClass:[IOSClass class]]) {
+    // IOSClass instances are always constant.
+    return YES;
+  }
   NSUInteger retainCount = [obj retainCount];
   return retainCount == UINT_MAX || retainCount == INT_MAX;
 }
diff --git a/j2objc/jre_emul/Classes/NSString+JavaString.m b/j2objc/jre_emul/Classes/NSString+JavaString.m
index 3adad8b..af0cc07 100644
--- a/j2objc/jre_emul/Classes/NSString+JavaString.m
+++ b/j2objc/jre_emul/Classes/NSString+JavaString.m
@@ -931,6 +931,10 @@ NSString *NSString_formatWithJavaUtilLocale_withNSString_withNSObjectArray_(
   return [self isEqualToString:[sb description]];
 }

+- (IOSClass *)getClass {
+  return NSString_class_();
+}
+
 jint javaStringHashCode(NSString *string) {
   static const char *hashKey = "__JAVA_STRING_HASH_CODE_KEY__";
   id cachedHash = objc_getAssociatedObject(string, hashKey);
diff --git a/j2objc/jre_emul/Tests/SerializationTest.java b/j2objc/jre_emul/Tests/SerializationTest.java
index 2077fc8..4d9a531 100644
--- a/j2objc/jre_emul/Tests/SerializationTest.java
+++ b/j2objc/jre_emul/Tests/SerializationTest.java
@@ -51,6 +51,9 @@ public class SerializationTest extends TestCase {
     }
   }

+  static class SerializableClass implements Serializable {}
+  static class NotSerializableClass {}
+
   @Override
   protected void tearDown() throws Exception {
     //new File(TEST_FILE_NAME).delete();
@@ -94,4 +97,31 @@ public class SerializationTest extends TestCase {
     ois.close();
     assertTrue("arrays not equal", Arrays.equals(names, result));
   }
+
+  // Regression test for https://github.com/google/j2objc/issues/496.
+  public void testSerializingObjectClass() throws Exception {
+    String path = "/tmp/a.object";
+    FileOutputStream fileOut = new FileOutputStream(path);
+    new ObjectOutputStream(fileOut).writeObject(Object.class);
+    FileInputStream fileIn = new FileInputStream(path);
+    new ObjectInputStream(fileIn).readObject();
+  }
+
+  // Regression test for https://github.com/google/j2objc/issues/496.
+  public void testSerializingSerializableClass() throws Exception {
+    String path = "/tmp/b.object";
+    FileOutputStream fileOut = new FileOutputStream(path);
+    new ObjectOutputStream(fileOut).writeObject(SerializableClass.class);
+    FileInputStream fileIn = new FileInputStream(path);
+    new ObjectInputStream(fileIn).readObject();
+  }
+
+  // Regression test for https://github.com/google/j2objc/issues/496.
+  public void testSerializingNotSerializableClass() throws Exception {
+    String path = "/tmp/c.object";
+    FileOutputStream fileOut = new FileOutputStream(path);
+    new ObjectOutputStream(fileOut).writeObject(NotSerializableClass.class);
+    FileInputStream fileIn = new FileInputStream(path);
+    new ObjectInputStream(fileIn).readObject();
+  }
 }
--
2.2.2
```

**Comment by tomball** [comment]
Fixed in 0.9.7.
